server/routes_test.go

```diff
@@ -608,3 +608,85 @@ func TestPutVersions(t *testing.T) {
 	}
 
 }
+
+func TestAuthorizeRequest(t *testing.T) {
+	u := auth.NewUser("test", []string{"returntrue"})
+	a1 := knox.Access{ID: "test", AccessType: knox.Read, Type: knox.User}
+
+	key := &knox.Key{
+		ID:  "test",
+		ACL: knox.ACL{a1},
+	}
+
+	mockCallbackTrue := func(input knox.AccessCallbackInput) (bool, error) {
+		return true, nil
+	}
+	mockCallbackFalse := func(input knox.AccessCallbackInput) (bool, error) {
+		return false, nil
+	}
+
+	// We will ignore the key's ACL because we are only testing the callback,
+	// which should return true if the AccessType is Write.
+	mockCallbackFalseWithValidInput := func(input knox.AccessCallbackInput) (bool, error) {
+		return input.AccessType == knox.Write && input.Key.ACL[0].ID == "test" && input.Key.ACL[0].Type == knox.User, nil
+	}
+
+	// Mock callback that panics
+	mockCallbackPanic := func(input knox.AccessCallbackInput) (bool, error) {
+		panic("intentional panic")
+	}
+
+	var (
+		authorized bool
+		err        error
+	)
+
+	defer func() {
+		SetAccessCallback(nil) // Resetting the callback
+	}()
+
+	SetAccessCallback(mockCallbackTrue)
+	authorized, err = authorizeRequest(key, u, knox.Write)
+	if err != nil {
+		t.Fatalf("Expected authorizeRequest to return nil error when accessCallback returns true, got %v", err)
+	}
+	if !authorized {
+		t.Fatal("Expected authorizeRequest to return true when accessCallback returns true")
+	}
+
+	SetAccessCallback(mockCallbackFalseWithValidInput)
+	authorized, err = authorizeRequest(key, u, knox.Write)
+	if err != nil {
+		t.Fatalf("Expected authorizeRequest to return nil error when accessCallback returns false with valid input, got %v", err)
+	}
+	if !authorized {
+		t.Fatal("Expected authorizeRequest to return true when accessCallback returns false with valid input")
+	}
+
+	SetAccessCallback(mockCallbackFalse)
+	authorized, err = authorizeRequest(key, u, knox.Write)
+	if err != nil {
+		t.Fatalf("Expected authorizeRequest to return nil error when accessCallback returns false, got %v", err)
+	}
+	if authorized {
+		t.Fatal("Expected authorizeRequest to return false when accessCallback returns false")
+	}
+
+	SetAccessCallback(nil)
+	authorized, err = authorizeRequest(key, u, knox.Write)
+	if err != nil {
+		t.Fatalf("Expected authorizeRequest to return nil error when accessCallback is nil, got %v", err)
+	}
+	if authorized {
+		t.Fatal("Expected authorizeRequest to return false when accessCallback is nil")
+	}
+
+	SetAccessCallback(mockCallbackPanic)
+	authorized, err = authorizeRequest(key, u, knox.Write)
+	if err == nil {
+		t.Fatal("Expected authorizeRequest to return error when accessCallback panics")
+	}
+	if authorized {
+		t.Fatal("Expected authorizeRequest to return false when accessCallback panics")
+	}
```

Comment:
nit:

you can create a testCase datastructure
```
struct testCase {
  name string
  callbackImpl func(input knox.AccessCallbackInput) (bool, error)
  testInput ...
  expectedOutput ...
}
```
then you can invoke in a loop like (pseudocode):
```
for case in cases:
  t.Run(case.name, func(t *testing.T) {
    shared test logic
  })
```

Reply:
good advice!